contracts/contracts/strategies/sonic/SonicValidatorDelegator.sol

```diff
@@ -0,0 +1,352 @@
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(withdrawal.validatorId > 0, "Invalid withdrawId");
+        return withdrawal.undelegatedAmount == 0;
+    }
+
+    /**
+     * @notice Restake any pending validator rewards for all supported validators
+     * @param validatorIds List of Sonic validator IDs to restake rewards
+     */
+    function restakeRewards(uint256[] calldata validatorIds)
+        external
+        nonReentrant
+    {
+        for (uint256 i = 0; i < validatorIds.length; ++i) {
+            uint256 rewards = sfc.pendingRewards(
+                address(this),
+                validatorIds[i]
+            );
+
+            if (rewards > 0) {
+                sfc.restakeRewards(validatorIds[i]);
+            }
+        }
+
+        // The SFC contract will emit Delegated and RestakedRewards events.
+        // The checkBalance function should not change as the pending rewards will moved to the staked amount.
+    }
+
+    /**
+     * @notice To receive native S from SFC and Wrapped Sonic (wS)
+     *
+     * @dev This does not prevent donating S tokens to the contract
+     * as wrappedSonic has a `withdrawTo` function where a third party
+     * owner of wrappedSonic can withdraw to this contract.
+     */
+    receive() external payable {
+        require(
+            msg.sender == address(sfc) || msg.sender == wrappedSonic,
+            "S not from allowed contracts"
+        );
+    }
+
+    /***************************************
+                Admin functions
+    ****************************************/
+
+    /// @notice Set the address of the Registrator which can delegate, undelegate and withdraw
+    function setRegistrator(address _address) external onlyGovernor {
+        validatorRegistrator = _address;
+        emit RegistratorChanged(_address);
+    }
+
+    /// @notice Set the default validatorId to delegate to on deposit
+    function setDefaultValidatorId(uint256 validatorId)
+        external
+        onlyRegistratorOrStrategist
+    {
+        require(isSupportedValidator(validatorId), "Validator not supported");
+        defaultValidatorId = validatorId;
+        emit DefaultValidatorIdChanged(validatorId);
+    }
+
+    /// @notice Allows a validator to be delegated to by the Registrator
+    function supportValidator(uint256 validatorId) external onlyGovernor {
+        require(
+            !isSupportedValidator(validatorId),
+            "Validator already supported"
+        );
+
+        supportedValidators.push(validatorId);
+
+        emit SupportedValidator(validatorId);
+    }
+
+    /// @notice Removes a validator from the supported list.
+    /// Unsupported validators can still be undelegated from, withdrawn from and rewards collected.
+    function unsupportValidator(uint256 validatorId) external onlyGovernor {
+        require(isSupportedValidator(validatorId), "Validator not supported");
+        require(
```

Comment:
❓would it be easier to kick off a withdraw here rather than validate stake is zero. The public can move rewards to stakes, so it's going to be very hard to empty a validator particularly when there's a timelock involved.

Reply:
good point, un-delegating everything when unsupporting a validator makes sense. Fixed here: https://github.com/OriginProtocol/origin-dollar/pull/2335/commits/51e25e719ae577017893962b2226e6bb0dce7d40